File: core/clientLaya/game/src/commonGame/tool/func/PlayerItemDicContainerTool.ts
namespace Shine
{
    /** 数据注册类 */
    export class PlayerItemDicContainerTool extends ItemDicContainerTool implements IPlayerFuncTool
    {
        public me:Player;

        constructor(funcID:number)
        {
            super(funcID);
        }

        public setMe(player:Player):void
        {
            this.me=player;
        }

        public createItemByType(type:number)
        {
            return this.me.bag.createItemByType(type);
        }

        public releaseItem(data:ItemData):void
        {
             this.me.bag.releaseItem(data);
        }

        public onItemAdd(index:number,data:ItemData,num:number):void
        {
            super.onItemAdd(index,data,num);
            //new number[]{_funcID,index}
            var arr:number[] = [this._funcID,index];

            this.me.dispatch(GameEventType.FuncItemContainerRefreshGrid,arr);

            if(this._funcID==FunctionType.MainBag)
            {
                 this.me.dispatch(GameEventType.MainBagRefreshGrid,index);
                 this.me.dispatch(GameEventType.MainBagItemChange,data.id);
            }
        }

        public onItemRemove(index:number,data:ItemData,num:number):void
        {
            super.onItemRemove(index,data,num);
            var arr:number[] = [this._funcID,index];
            this.me.dispatch(GameEventType.FuncItemContainerRefreshGrid,arr);

            if(this._funcID==FunctionType.MainBag)
            {
                 this.me.dispatch(GameEventType.MainBagRefreshGrid,index);
                 this.me.dispatch(GameEventType.MainBagItemChange,data.id);
            }
        }

        public onChanged():void
        {
            super.onChanged();

             this.me.dispatch(GameEventType.FuncItemContainerChange,this._funcID);

            if(this._funcID==FunctionType.MainBag)
            {
                 this.me.dispatch(GameEventType.MainBagChange);
            }

            // this.me.bag.printBag();
        }

        public onRefreshAll():void
        {
             this.me.dispatch(GameEventType.FuncItemContainerRefreshAll,this._funcID);

            if(this._funcID==FunctionType.MainBag)
            {
                 this.me.dispatch(GameEventType.MainBagRefreshAll);
            }

            super.onRefreshAll();
        }

        public checkCanUseItem(data:ItemData,num:number,arg:UseItemArgData,needNotice:boolean):boolean
        {
            return  this.me.bag.checkItemUseConditions(data,num,arg,needNotice);
        }

        public toSendUseItem(data:ItemData,index:number,num:number,arg:UseItemArgData):void
        {
             this.me.send(FuncUseItemRequest.createFuncUseItem(this._funcID,index,num,data.id,arg));
        }

        public toUseItem(data:ItemData,num:number,arg:UseItemArgData):void
        {
             this.me.bag.toUseItem(data,num,arg);
        }

        /** 执行使用物品通过id */
        public toUseItemId(id:number,num:number,arg:UseItemArgData):void
        {
            this.me.bag.toUseItemId(id,num,arg);
        }

        /** 服务器返回使用物品结果 */
        public doUseItemResult(id:number,num:number,result:boolean):void
        {
            if(this._funcID==FunctionType.MainBag)
            {
                 this.me.bag.useItemResult(id,num,result);
            }
        }

        public onRedPointChange():void
        {
            if(this._funcID==FunctionType.MainBag)
            {
               // GameC.redPoint.refreshOne(RedPointType.Bag);
            }
        }

        public onRedPointRemoved(index:number):void
        {
            super.onRedPointRemoved(index);

            //发送服务器
             this.me.send(FuncItemRemoveRedPointRequest.createFuncItemRemoveRedPoint(this._funcID,index));

            var arr:number[] = [this._funcID,index];
            //更新单格
             this.me.dispatch(GameEventType.FuncItemContainerRefreshGrid,arr);

            if(this._funcID==FunctionType.MainBag)
            {
                 this.me.dispatch(GameEventType.MainBagRefreshGrid,index);
            }
        }

        public onAddItemNotice(list:SList<ItemData>,way:number):void
        {
            var param:SMap<string,Object> =new SMap<string,Object>();
            param.put("list",list);
            param.put("way",way);

            this.me.dispatch(GameEventType.AddItemNotice,param);
        }
    }
}

File: core/client/game/src/commonGame/tool/func/PlayerItemDicContainerTool.cs
using System;
using ShineEngine;

/// <summary>
///
/// </summary>
public class PlayerItemDicContainerTool:ItemDicContainerTool,IPlayerFuncTool
{
	public Player me;

	public PlayerItemDicContainerTool(int funcID):base(funcID)
	{

	}

	public void setMe(Player player)
	{
		me=player;
	}

	protected override ItemData createItemByType(int type)
	{
		return me.bag.createItemByType(type);
	}

	protected override void releaseItem(ItemData data)
	{
		me.bag.releaseItem(data);
	}

	protected override void onItemAdd(int index,ItemData data,int num,int way)
	{
		base.onItemAdd(index,data,num,way);

		me.dispatch(GameEventType.FuncItemContainerRefreshGrid,new int[]{_funcID,index});

		if(_funcID==FunctionType.MainBag)
		{
			me.dispatch(GameEventType.MainBagRefreshGrid,index);
			me.dispatch(GameEventType.MainBagItemChange,data.id);
		}
	}

	protected override void onItemRemove(int index,ItemData data,int num,int way)
	{
		base.onItemRemove(index,data,num,way);

		me.dispatch(GameEventType.FuncItemContainerRefreshGrid,new int[]{_funcID,index});

		if(_funcID==FunctionType.MainBag)
		{
			me.dispatch(GameEventType.MainBagRefreshGrid,index);
			me.dispatch(GameEventType.MainBagItemChange,data.id);
		}
	}

	protected override void onChanged()
	{
		base.onChanged();

		me.dispatch(GameEventType.FuncItemContainerChange,_funcID);

		if(_funcID==FunctionType.MainBag)
		{
			me.dispatch(GameEventType.MainBagChange);
		}

		//me.bag.printBag();
	}

	protected override void onRefreshAll()
	{
		me.dispatch(GameEventType.FuncItemContainerRefreshAll,_funcID);

		if(_funcID==FunctionType.MainBag)
		{
			me.dispatch(GameEventType.MainBagRefreshAll);
		}

		base.onRefreshAll();
	}

	public override bool checkCanUseItem(ItemData data,int num,UseItemArgData arg,bool needNotice)
	{
		return me.bag.checkItemUseConditions(data,num,arg,needNotice);
	}

	protected override void toSendUseItem(ItemData data,int index,int num,UseItemArgData arg)
	{
		me.send(FuncUseItemRequest.create(_funcID,index,num,data.id,arg));
	}

	protected override void toUseItem(ItemData data,int num,UseItemArgData arg)
	{
		me.bag.toUseItem(data,num,arg);
	}

	protected override void toUseItem(int id,int num,UseItemArgData arg)
	{
		me.bag.toUseItem(id,num,arg);
	}

	/** 服务器返回使用物品结果 */
	protected override void doUseItemResult(int id,int num,Boolean result)
	{
		if(_funcID==FunctionType.MainBag)
		{
			me.bag.useItemResult(id,num,result);
		}
	}

	protected override void onRedPointChange()
	{
		if(_funcID==FunctionType.MainBag)
		{
			GameC.redPoint.refreshOne(RedPointType.Bag);
		}
	}

	protected override void onRedPointRemoved(int index)
	{
		base.onRedPointRemoved(index);

		//发送服务器
		me.send(FuncItemRemoveRedPointRequest.create(_funcID,index));

		//更新单格
		me.dispatch(GameEventType.FuncItemContainerRefreshGrid,new int[]{_funcID,index});

		if(_funcID==FunctionType.MainBag)
		{
			me.dispatch(GameEventType.MainBagRefreshGrid,index);
		}
	}

	protected override void onAddItemNotice(SList<ItemData> list,int way)
	{
		SMap<string,object> param=new SMap<string, object>();
		param.put("list",list);
		param.put("way",way);

		me.dispatch(GameEventType.AddItemNotice,param);
	}
}